scripts: add MACHO lazy bindings check to security-check.py

fanquake opened the pull request:
This is a slightly belated follow up to #17686 and some discussion with Cory. It's not entirely clear if we should make this change due to the way the macOS dynamic loader appears to work. However I'm opening this for some discussion. Also related to #17768.

#### Issue:
[`LD64`](https://opensource.apple.com/source/ld64/) doesn't set the [MH_BINDATLOAD](https://opensource.apple.com/source/xnu/xnu-6153.11.26/EXTERNAL_HEADERS/mach-o/loader.h.auto.html) bit in the header of MACHO executables, when building with `-bind_at_load`. This is in contradiction to the [documentation](https://opensource.apple.com/source/ld64/ld64-450.3/doc/man/man1/ld.1.auto.html):
```bash
-bind_at_load
     Sets a bit in the mach header of the resulting binary which tells dyld to
     bind all symbols when the binary is loaded, rather than lazily.
```

The [`ld` in Apples cctools](https://opensource.apple.com/source/cctools/cctools-927.0.2/ld/layout.c.auto.html) does set the bit, however the [cctools-port](https://github.com/tpoechtrager/cctools-port/) that we use for release builds, bundles `LD64`.

However; even if the linker hasn't set that bit, the dynamic loader ([`dyld`](https://opensource.apple.com/source/dyld/)) doesn't seem to ever check for it, and from what I understand, it looks at a different part of the header when determining whether to lazily load symbols. 

Note that our release binaries are currently working as expected, and no lazy loading occurs.

#### Example:

Using a small program, we can observe the behaviour of the dynamic loader.

Conducted using:
```bash
clang++ --version
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin18.7.0

ld -v
@(#)PROGRAM:ld  PROJECT:ld64-530
BUILD 18:57:17 Dec 13 2019
LTO support using: LLVM version 11.0.0, (clang-1100.0.33.17) (static support for 23, runtime is 23)
TAPI support using: Apple TAPI version 11.0.0 (tapi-1100.0.11)
```

```cpp
#include <iostream>
int main() {
	std::cout << "Hello World!\n";
	return 0;
}
```

Compile and check the MACHO header:
```bash
clang++ test.cpp -o test
otool -vh test
...
Mach header
      magic cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
MH_MAGIC_64  X86_64        ALL LIB64     EXECUTE    16       1424   NOUNDEFS DYLDLINK TWOLEVEL WEAK_DEFINES BINDS_TO_WEAK PIE

# Run and dump dynamic loader bindings:
DYLD_PRINT_BINDINGS=1 DYLD_PRINT_TO_FILE=no_bind.txt ./test
Hello World!
```

Recompile with `-bind_at_load`. Note still no `BINDATLOAD` flag:
```bash
clang++ test.cpp -o test -Wl,-bind_at_load
otool -vh test
Mach header
      magic cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
MH_MAGIC_64  X86_64        ALL LIB64     EXECUTE    16       1424   NOUNDEFS DYLDLINK TWOLEVEL WEAK_DEFINES BINDS_TO_WEAK PIE
...
DYLD_PRINT_BINDINGS=1 DYLD_PRINT_TO_FILE=bind.txt ./test
Hello World!
```

If we diff the outputs, you can see that `dyld` doesn't perform any lazy bindings when the binary is compiled with `-bind_at_load`, even if the `BINDATLOAD` flag is not set:
```diff
@@ -1,11 +1,27 @@
+dyld: bind: test:0x103EDF030 = libc++.1.dylib:__ZNKSt3__16locale9use_facetERNS0_2idE, *0x103EDF030 = 0x7FFF70C9FA58
+dyld: bind: test:0x103EDF038 = libc++.1.dylib:__ZNKSt3__18ios_base6getlocEv, *0x103EDF038 = 0x7FFF70CA12C2
+dyld: bind: test:0x103EDF068 = libc++.1.dylib:__ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEE6sentryC1ERS3_, *0x103EDF068 = 0x7FFF70CA12B6
+dyld: bind: test:0x103EDF070 = libc++.1.dylib:__ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEE6sentryD1Ev, *0x103EDF070 = 0x7FFF70CA1528
+dyld: bind: test:0x103EDF080 = libc++.1.dylib:__ZNSt3__16localeD1Ev, *0x103EDF080 = 0x7FFF70C9FAE6
<trim>
-dyld: lazy bind: test:0x10D4AC0C8 = libsystem_platform.dylib:_strlen, *0x10D4AC0C8 = 0x7FFF73C5C6E0
-dyld: lazy bind: test:0x10D4AC068 = libc++.1.dylib:__ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEE6sentryC1ERS3_, *0x10D4AC068 = 0x7FFF70CA12B6
-dyld: lazy bind: test:0x10D4AC038 = libc++.1.dylib:__ZNKSt3__18ios_base6getlocEv, *0x10D4AC038 = 0x7FFF70CA12C2
-dyld: lazy bind: test:0x10D4AC030 = libc++.1.dylib:__ZNKSt3__16locale9use_facetERNS0_2idE, *0x10D4AC030 = 0x7FFF70C9FA58
-dyld: lazy bind: test:0x10D4AC080 = libc++.1.dylib:__ZNSt3__16localeD1Ev, *0x10D4AC080 = 0x7FFF70C9FAE6
-dyld: lazy bind: test:0x10D4AC070 = libc++.1.dylib:__ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEE6sentryD1Ev, *0x10D4AC070 = 0x7FFF70CA1528
```

Note: `dyld` also has a `DYLD_BIND_AT_LAUNCH=1` environment variable, that when set, will force any lazy bindings to be non-lazy:
```bash
dyld: forced lazy bind: test:0x10BEC8068 = libc++.1.dylib:__ZNSt3__113basic_ostream
```

#### Thoughts:
After looking at the dyld source, I can't find any checks for `MH_BINDATLOAD`. You can see the flags it does check for, such as MH_PIE or MH_BIND_TO_WEAK [here](https://opensource.apple.com/source/dyld/dyld-732.8/src/ImageLoaderMachO.cpp.auto.html).

It seems that the lazy binding of any symbols depends on whether or not [lazy_bind_size](https://opensource.apple.com/source/xnu/xnu-6153.11.26/EXTERNAL_HEADERS/mach-o/loader.h.auto.html) from the `LC_DYLD_INFO_ONLY` load command is > 0. Which was mentioned in [#17686](https://github.com/bitcoin/bitcoin/pull/17686#issue-350216254).

#### Changes:
This PR is one of [Corys commits](https://github.com/theuni/bitcoin/commit/7b6ba26178d2754568a1308d3d44e038e9ebf450), that I've rebased and modified to make build. I've also included an addition to the `security-check.py` script to check for the flag.

However, given the above, I'm not entirely sure this patch is the correct approach. If the linker no-longer inserts it, and the dynamic loader doesn't look for it, there might be little benefit to setting it. Or, maybe this is an oversight from Apple and needs some upstream discussion. Looking for some thoughts / Concept ACK/NACK. 

One alternate approach we could take is to drop the patch and modify security-check.py to look for `lazy_bind_size` == 0 in the `LC_DYLD_INFO_ONLY` load command, using `otool -l`.

fanquake committed:
scripts: add MACHO lazy bindings check to security-check.py

laanwj commented:
@fanquake and me had some discussion about this on IRC. The conclusion was that it's probably better to ignore the `MH_BINDATLOAD` flag like MacOS' static linker does. Instead, to adopt the security check for the change(s) that `-Wl,-bind_at_load` actually makes and is used.

Also, it would make sense to get Apple to clarify the documentation on this. It happens more often that header bits simply go unused in new versions in favor of new mechanisms, but it would have avoided a lot of hassle if this had been properly documented.

fanquake commented:
> Also, it would make sense to get Apple to clarify the documentation on this

I've reached out to a few people at Apple. So hopefully we'll get some clarification.

laanwj commented:
Great!

fanquake commented:
Received this reply from [Nick Kledzik](https://opensource.apple.com/source/ld64/ld64-450.3/doc/design/linker.html.auto.html) (ld / dyld at Apple):
> Michael,

> Traditionally, there is two kinds of binding.  Binds which must be done at load time before any code is run, and binds which can be done lazily.  Lazy binds are only used for procedural interfaces.  That is, the symbol is an external function call.  The linker tool and dyld conspire to have lazying bind work by having the actual pointer bound to a helper which on first call looks up and set the pointer to the real implementation.  This lazy symbol optimization made sense if looking up symbols took a long time and the program was not going to call them immediately. With dyld3 all symbol lookups are done ahead of time and cached, so lazy bind makes no sense.

> The semantics of the -bind_at_load option is omit that optimization.  That is, have all binding done at launch.  Originally that was done by setting the MH_BINDATLOAD bit in the mach_header and have dyld bind all lazy pointer early.

> A few years back we changed to instead have the static linker generate the binary without lazy binding - just up front binding for all symbols.  You can see that in this example:

```bash
# regular build (has lazy binding)
[/tmp]> xcrun -sdk macosx.internal clang main.c
[/tmp]> xcrun dyldinfo -lazy_bind -bind a.out
bind information:
segment section          address        type    addend dylib            symbol
__DATA_CONST __got            0x100001000    pointer      0 libSystem        dyld_stub_binder
lazy binding information (from lazy_bind part of dyld info):
segment section          address    index  dylib                        symbol
__DATA       __la_symbol_ptr  0x100002000 0x0000           libSystem        _printf
```

```bash
# -bind_at_load build (no lazy binding)
[/tmp]> xcrun -sdk macosx.internal clang main.c -bind_at_load
[/tmp]> xcrun dyldinfo -lazy_bind -bind a.out
bind information:
segment section          address        type    addend dylib              symbol
__DATA       __la_symbol_ptr  0x100002000    pointer      0 libSystem        _printf
__DATA_CONST __got            0x100001000    pointer      0 libSystem        dyld_stub_binder
no compressed lazy binding info
```

laanwj commented:
Thanks for asking for clarification.

So from what I understand essentially this means that there is no point in forcing `MH_BINDATLOAD` because it's no longer necessary?

I think this means we don't need this and can close it?

fanquake commented:
> I think this means we don't need this and can close it?

I've asked Cory to follow up here now that we have some more information. I think even if we didn't want to use `MH_BINDATLOAD` we can at least add a related check to `security-check.py`.

theuni commented:
@fanquake Thanks for pinging someone at Apple!

I think I understand now, and that makes sense. I wasn't aware of the old faux-static linkage, but the flag makes sense in that context. Sounds like it produced the same behavior that a normal link and DYLD_BIND_AT_LAUNCH at runtime would now.

That confirms (to me, anyway) that we're doing the right (extra paranoid) thing by turning on bind_at_load and sacrificing startup time to avoid stub binding that obfuscates what's happening at runtime, which proved to be a real issue recently: https://github.com/bitcoin-core/secp256k1/issues/674#issuecomment-548452744

It also explains why the dyloader doesn't bother checking for the flag at runtime, so there's no need for us to set the bit.

It _also_ confirms that your thoughts on checks in #17686 were spot on:
>"You can check the binaries using otool -l, and looking for the LC_DYLD_INFO_ONLY section;
> lazy_bind_off and lazy_bind_size should both be 0."

So, let's go for that instead, unless it's already been merged as part of another PR?

Kudos to @fanquake for looking so deeply into this, this was some really great detective work.

fanquake commented:
> So, let's go for that instead,

Thanks for following up. I've added a different test to security-check.py, which checks the `otool -l` output, and dropped the other changes.

theuni commented:
ACK 5ca90f8b598978437340bb8467f527b9edfb2bbf
